### ISSUE TITLE
Separate LFPy forward models from LFPy itself

### ISSUE THREAD

**Issue by espenhgn** [LFPy/LFPy#187]
It would be beneficial to split the different forward models in LFPy from the main LFPy module, and put them in a separate repository/module. This would allow us and others to utilize the forward models in their own projects without dependency on NEURON, 
and it would make future integration with e.g., Arbor easier (https://github.com/arbor-sim/arbor/issues/1036, https://github.com/multiscale-cosim/EBRAINS-cosim/issues/18).

My proposal would be that each forward model is reimplemented as a subclass of `LinearModel`, which should take a `GenericCell` object instance as input:
```
class GenericCell(object):
    def __init__(self, **kwargs):
        '''
        Generic cell class with compartment geometry information
        
        Parameters
        ----------
        xstart, ystart, zstart : ndarray
            arrays of length totnsegs with start (x,y,z) coordinate of segments
            in units of um
        xmid, ymid, zmid : ndarray
            midpoint coordinates of segments
        xend, yend, zend : ndarray
            endpoint coordinateso of segments
        diam : ndarray
            array of length totnsegs with segment diameters in units of um
        area : ndarray
            array of segment surface areas in units of um^2
        '''
        # set attributes
        self.xstart = xstart
        # derived attributes
        self.totnsegs = xstart.size 
        ...

class LinearModel(object):
    def __init__(self, **kwargs):
        '''
        Parameters
        ------------
        cell: GenericCell or LFPy.Cell like instance
        input: str
            'imem' or 'vmem' or 'ipas' or ...
        arg: ...
        '''
            self.cell = cell
            self.input = input
            
            
    def get_linear_transform(self):
        '''
        Return transformation **M** that linearly transforms an array of <input> values **I**
        into some signal **S** given as **S=MI**

        Here's an example mapping that just returns <input>
        '''
        return np.eye(self.cell.totnsegs)
```

In an LFPy simulation a list of object instances inherited from `<LinearModel>` for electrodes, dipole moments and so forth could be supplied to `LFPy.Cell.simulate()` which would then call the `get_linear_transform` for each 
precompute the mappings and use them. We need a neat way of specifying what it should depend on though, like transmembrane currents, voltages etc.:
```
import LFPy
from LFPyLinearModels import LinearModel
# create objects
cell = LFPy.Cell(**cellargs)
linearmodel = LinearModel(**linearmodelargs)
# instrument cell with synapses etc.
...
# run model
cell.simulate(transforms=[linearmodel], **simulateargs)
# get data
cell.tvec  # simulation time points
linearmodel.data  # predicted values from linear model during the course of the simulation.
```

@torbjone, what do you think?

**Comment by torbjone** [comment]
Do you think this would be beneficial also for LFPy itself, or only for potential integration with non-NEURON based software? If the former is true, perhaps we should have a brief meeting and discuss the needed modifications while looking at the code? While if the latter is true, maybe we should wait until we are sure this is the solution we will go for with regards to integration with Arbor?

**Comment by espenhgn** [comment]
Hi @torbjone, 
I think for LFPy it would be better to set up a uniform method to make predictions and save output to memory/file using the different models. Right now the `Cell.simulate` method takes both arguments `electrode` (`RecExtElectrode` object or list thereof), `rec_current_dipole_moment` (bool) and `dotprodcoeffs` (list of linear mappings) and as a result the `_run_simulation_with_electrode` extension is a mess that is difficult to optimize. Those above arguments can be turned into a single argument with ease; accepting just as a list of such plugins. 

As for other changes, `RecExtElectrode` and the `lfpcalc.calc_lfp*` methods already take a `Cell` as the main parameter and doesn't rely on NEURON so these would be trivial to relocate. Another class that provides a mapping for current dipole moment would be easy to implement as well. For ground-truth CSDs (#185) I would also prefer to define a similar class to map currents to volume densities. Some unit tests will have to be updated to reflect the above changes. 

With regards to Arbor, it would be useful to get @halfflat's input I believe.

**Comment by halfflat** [comment]
Regarding Arbor's view of things, my ideas are still rather inchoate. But if it's always going to be a matter of computing linear responses to sampled electrical data, then my thinking is that the cleanest interface would follow a structure such as:
1. Model (cells, networking, etc.) is set up in Arbor.
2. Arbor provides the required metadata, which I believe would comprise point or piecewise-linear cable segment geometry that corresponds to the sampling or discretization scenario.
3. LFPy provides the response matrix (in some representation, which may well be distributed) in turn, and gives that to a special handler in Arbor (this bit would be new).
4. When the simulation is run, we apply the response matrix and either collate the results for consumption at the end of the simulation, or stream it via some sort of callback process.

Will everything that is required be able to be modelled in this sort of fashion?

**Comment by espenhgn** [comment]
Hi @halfflat , 
Thanks for your input. I discussed this matter with @torbjone earlier today, and I believe we can make this work. 
1. We need access to cell geometry: basically the start and endpoint coordinate of each compartment, as well as the compartment diameter. It appears as Arbor treats compartments as straight, variable diameter cylinders, while we have assumed in the past that compartments in NEURON are straight cylinders with constant diameter and constant transmembrane current density per length unit/surface area. As a first approximation I will assume we can treat compartments from Arbor the same way. LFPy presently uses [µm] for spatial units. 
2. We can provide the response matrix that transforms a vector (or matrix for multiple timesteps) of currents into an extracellular potential. Right now we would assume that the unit of compartment currents are similar to those of NEURON with use of CVode.use_fast_imem(), i.e., [nA] (https://www.neuron.yale.edu/neuron/static/new_doc/simctrl/cvode.html#CVode.use_fast_imem). 
If such matrices can be computed on the Python level and be pushed to the C++ level that would be great, but I can imagine that we also can convert methods to Cython if we have to. I believe that would allow usage of the methods directly from C++. 
3. The response matrices can be computed and utilized locally on each process for the cells/compartments on that process. Summing of all contributions can be performed when the simulation is finished or during simulation but I suspect the latter is less efficient. 
4. Having options for saving output to file or just keep things in memory would be good. 

With this I believe our short term strategy is to put different forward-models in a separate python module, and define a uniform interface for it in order to utilize the different forward-models included in the module.

**Comment by halfflat** [comment]
This all sounds very reasonable!

There is one (minor) complication, in that typically, for Arbor, the discretization of the dynamics into compartments (we call them CVs in the documentation to distinguish them from NEURON compartments — they're not quite the same thing) doesn't line up with the discretization of the morphology into piecewise-linear segments.

Consequently, any one sample of a trans-membrane current on the cell corresponds to the current over multiple cylinders (or conical frustra as the case may be); depending on the discretization, they might even constitute a forked sub-tree of the morphology. We can apportion the current across each segment proportionally by area, and sum against the response for each of these segments, but it would be more efficient to do it one step, and have the response for a CV current be the area-weighted sum of the response from each segment.

**Comment by espenhgn** [comment]
Hi @halfflat. Doing a weighting of contributions per area for each CV sounds doable. Thanks for the clarification!

**Comment by espenhgn** [comment]
I created a new repo/project at https://github.com/LFPy/lfpy_forward_models where I'll start implementing things.

**Comment by espenhgn** [comment]
Just a little status update. The basic set up is in place (python module, test module, continuous integration, lint checks, couple of examples, etc.). 
I need a "better" module name though. Any suggestions? lfpylib? lfpykit? lfpytools?

I'm changing things slightly from above. The way present LFPy stores Cell geometry information is convoluted (for historical reasons), where the following attributes are set: `xstart, xmid, xend, ystart, ... , totnsegs, length, area`. The `CellGeometry` class ( https://github.com/LFPy/lfpy_forward_models/blob/4342023b635d3d89d72851a8854b026b2e261016/lfpy_forward_models/cell.py#L19) should simplify that to 
```
    Parameters
    ----------
    x: ndarray of floats
        shape (n_seg x 2) array of start- and end-point coordinates of
        each segment along x-axis in units of [um]
    y: ndarray of floats
        shape (n_seg x 2) array of start- and end-point coordinates of
        each segment along y-axis in units of [um]
    z: ndarray of floats
        shape (n_seg x 2) array of start- and end-point coordinates of
        each segment along z-axis in units of [um]
    d: ndarray of floats
        shape (n_seg, ) array of segment diameters in units of [um]
        (shape (n_seg, 2) for conical frusta to be implemented) 
```
Segment lengths and areas are calculated upon instantiation. This should be leaner and cleaner than the current implementation in LFPy, and I would adapt the different forward models accordingly (and incorporate into LFPy at some later point - I see no reason why not except breaking some people's model codes :-/ ). 

Different forward models will be (re)implemented in https://github.com/LFPy/lfpy_forward_models/blob/master/lfpy_forward_models/models.py. Right now only `CurrentDipoleMoment` does anything useful, as demonstrated in the examples. I intend that other models can be utilized in a similar manner. 

Grouping of 'segments' belonging to different CVs from Arbor will be implemented at some point. I'll assume one can iterate over coordinates/diameters of each CV.

**Comment by espenhgn** [comment]
Fixed by merging #226
